ext/mbstring/libmbfl/filters/mbfilter_utf16.c

```diff
@@ -514,10 +658,387 @@ static void mb_wchar_to_utf16le(uint32_t *in, size_t len, mb_convert_buf *buf, b
+					}
+				} while (n_words);
+			}
+		} while (len >= 8);
+
+		MB_CONVERT_BUF_STORE(buf, out, limit);
+	}
+
+	if (len) {
+		mb_wchar_to_utf16be_default(in, len, buf, end);
+	}
+}
+
+#ifdef ZEND_INTRIN_AVX2_FUNC_PROTO
+size_t mb_utf16le_to_wchar_avx2(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+#else
+static size_t mb_utf16le_to_wchar_avx2(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+#endif
+{
+	/* Most of this function is the same as `mb_utf16be_to_wchar_avx2`, above;
+	 * See it for more detailed code comments */
+
+	size_t len = *in_len;
+
+	if (len >= 32 && bufsize >= 16) {
+		unsigned char *p = *in;
+		uint32_t *out = buf;
+
+		const __m256i _f8 = _mm256_set1_epi16(0xF800);
+		const __m256i _d8 = _mm256_set1_epi16(0xD800);
+
+		do {
+			__m256i operand = _mm256_loadu_si256((__m256i*)p);
+
+			uint32_t surrogate_bitvec = _mm256_movemask_epi8(_mm256_cmpeq_epi16(_mm256_and_si256(operand, _f8), _d8));
+			if (surrogate_bitvec == 0) {
+				/* There are no surrogates among these 16 characters */
+				_mm256_storeu_si256((__m256i*)out, _mm256_cvtepu16_epi32(_mm256_castsi256_si128(operand)));
+				_mm256_storeu_si256((__m256i*)(out + 8), _mm256_cvtepu16_epi32(_mm256_extracti128_si256(operand, 1)));
+				out += 16;
+				bufsize -= 16;
+				p += sizeof(__m256i);
+				len -= sizeof(__m256i);
+			} else if ((surrogate_bitvec & 1) == 0) {
+				/* Some prefix of the current block is non-surrogates */
+				uint8_t n_chars = zend_ulong_ntz(surrogate_bitvec) >> 1;
+				_mm256_storeu_si256((__m256i*)out, _mm256_cvtepu16_epi32(_mm256_castsi256_si128(operand)));
+				if (n_chars > 8) {
+					_mm256_storeu_si256((__m256i*)(out + 8), _mm256_cvtepu16_epi32(_mm256_extracti128_si256(operand, 1)));
+				}
+				out += n_chars;
+				bufsize -= n_chars;
+				p += n_chars * 2;
+				len -= n_chars * 2;
+			} else {
+				/* Some prefix of the current block is (valid or invalid) surrogates */
+				surrogate_bitvec = ~surrogate_bitvec;
+				unsigned int n_chars = surrogate_bitvec ? zend_ulong_ntz(surrogate_bitvec) >> 1 : 16;
+				do {
+					unsigned char c1 = *p++;
+					unsigned char c2 = *p++;
+
+					if (c2 & 0x4 || len < 4) {
+						/* 2nd part of surrogate pair has come first OR string ended abruptly
+						 * after 1st part of surrogate pair */
+						*out++ = MBFL_BAD_INPUT;
+						bufsize--;
+						n_chars--;
+						len -= 2;
+						continue;
+					}
+
+					uint16_t n = (c2 << 8) | c1;
+					unsigned char c3 = *p++;
+					unsigned char c4 = *p++;
+
+					if ((c4 & 0xFC) == 0xDC) {
+						/* Valid surrogate pair */
+						uint16_t n2 = (c4 << 8) | c3;
+						*out++ = (((n & 0x3FF) << 10) | (n2 & 0x3FF)) + 0x10000;
+						bufsize--;
+						len -= 4;
+#ifdef PHP_HAVE_BUILTIN_USUB_OVERFLOW
+						if (__builtin_usub_overflow(n_chars, 2, &n_chars)) {
+							break;
+						}
+#else
+						n_chars -= 2;
+						if (n_chars == UINT_MAX) {
```

Comment:
n_chars can be uneven prior to the execution of `n_chars -= 2`, right? In that case, n_chars - 2 will not equal UINT_MAX even though an overflow occurred. This would be caught by the __builtin_usub_overflow however. I think you need `n_chars >= UINT_MAX - 1` ? (same for the equivalent code above)

Reply:
You are sharp! Thanks for reading the code in detail!

`n_chars` can't be zero here, because if it was, we wouldn't have gone around the loop another time. If `n_chars` is 2 or more, then subtracting 2 is safe. The only problem comes in if `n_chars` is 1. Then subtracting 2 should make it wrap around to `UINT_MAX`.

Does that make sense?

Reply:
Right indeed, thanks for clarifying!